File: packages/native-paging/ios/NativePager.swift
import ExpoModulesCore
import SwiftUI
@_exported import VoelPagingCore

public struct PagingOptions: Record {
  @Field public var pageSize: Int = 50
  @Field public var maxResidentItems: Int = 250
  public init() {}
}

struct PagingItemRecord: Record {
  @Field(.required) var id: String = ""
  @Field(.required) var value: [String: Any] = [:]
}

public struct PageDelivery: Record {
  @Field(.required) var id: String = ""
  @Field(.required) var items: [PagingItemRecord] = []
  @Field(.required) var total: Int = 0
  public init() {}
}

public struct PageCancellation: Record {
  @Field(.required) var id: String = ""
  public init() {}
}

/// Main-thread state owned by the native view, not an Expo shared object.
public final class NativePager<Value: AnyObject>: ObservableObject {
  @Published public private(set) var snapshot = PageSnapshot<Value>(
    items: [], refresh: .loading, prepend: .ready, append: .ready
  )
  private var session: PagingSession<Value>?
  private let decode: ([String: Any]) throws -> Value
  private let generation = UUID().uuidString
  private var pending: [String: Int32] = [:]

  public init(
    options: PagingOptions,
    decode: @escaping ([String: Any]) throws -> Value,
    onRequest: @escaping ([String: Any]) -> Void,
    onCancel: @escaping ([String: Any]) -> Void
  ) {
    self.decode = decode
    session = PagingSession(
      pageSize: Int32(options.pageSize),
      maxItems: Int32(options.maxResidentItems),
      onRequest: { [weak self] request in
        guard let self else { return }
        let id = "\(generation):\(request.id)"
        pending[id] = request.id
        onRequest(["id": id, "offset": request.offset, "limit": request.limit])
      },
      onCancel: { [weak self] requestId in
        guard let self else { return }
        let id = "\(generation):\(requestId.intValue)"
        pending.removeValue(forKey: id)
        onCancel(["id": id])
      },
      onSnapshot: { [weak self] snapshot in self?.snapshot = snapshot }
    )
  }

  public func start() { session?.start() }
  public func access(id: String) { session?.access(id: id) }
  public func retry() { session?.retry() }
  public func refresh() { session?.refresh() }
  public func resolve(_ page: PageDelivery) throws {
    guard let id = pending[page.id] else { return }
    let items = try page.items.map { PageItem<Value>(id: $0.id, value: try decode($0.value)) }
    pending.removeValue(forKey: page.id)
    session?.resolve(id: id, items: items, total: Int32(page.total))
  }
  public func reject(_ request: PageCancellation) {
    guard let id = pending.removeValue(forKey: request.id) else { return }
    session?.reject(id: id)
  }
  public func close() {
    session?.close()
    session = nil
    pending.removeAll()
  }

  deinit {
    // ARC can release view props off the main thread. Do not retain the view until
    // a network request finishes; session callbacks above deliberately capture weakly.
    let session = session
    DispatchQueue.main.async { session?.close() }
  }
}

File: packages/native-paging/android/src/main/java/app/voel/paging/NativePager.kt
package app.voel.paging

import app.voel.paging.core.PageItem
import app.voel.paging.core.PageSnapshot
import app.voel.paging.core.PageStatus
import app.voel.paging.core.PagingSession
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import expo.modules.kotlin.records.Required
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.io.Serializable
import java.util.UUID

class PagingOptions : Record {
    @Field val pageSize: Int = 50
    @Field val maxResidentItems: Int = 250
}

class PagingItemRecord : Record {
    @Field @Required val id: String = ""
    @Field @Required val value: Map<String, Any?> = emptyMap()
}

class PageDelivery : Record {
    @Field @Required val id: String = ""
    @Field @Required val items: List<PagingItemRecord> = emptyList()
    @Field @Required val total: Int = 0
}

class PageCancellation(@Field val id: String = "") : Record, Serializable

class PageRequestEvent(
    @Field val id: String = "",
    @Field val offset: Int = 0,
    @Field val limit: Int = 0,
) : Record, Serializable

/** Main-thread, view-owned state. No Expo shared object or JavaScript-managed lifetime. */
class NativePager<Value : Any>(
    options: PagingOptions,
    private val decode: (Map<String, Any?>) -> Value,
    onRequest: (PageRequestEvent) -> Unit,
    onCancel: (PageCancellation) -> Unit,
) {
    // A recreated native view must never accept a late reply from its predecessor.
    private val generation = UUID.randomUUID().toString()
    private val pending = mutableMapOf<String, Int>()
    private val current = MutableStateFlow(
        PageSnapshot<Value>(emptyList(), PageStatus.LOADING, PageStatus.READY, PageStatus.READY)
    )
    val snapshots = current.asStateFlow()
    private var session: PagingSession<Value>? = PagingSession(
        pageSize = options.pageSize,
        maxItems = options.maxResidentItems,
        onRequest = {
            val id = "$generation:${it.id}"
            pending[id] = it.id
            onRequest(PageRequestEvent(id, it.offset, it.limit))
        },
        onCancel = {
            val id = "$generation:$it"
            pending.remove(id)
            onCancel(PageCancellation(id))
        },
        onSnapshot = { current.value = it },
    )

    fun start() { session?.start() }
    fun access(id: String) { session?.access(id) }
    fun retry() { session?.retry() }
    fun resolve(page: PageDelivery) {
        val id = pending[page.id] ?: return
        // Decode atomically; command failures reach JS and reject this request.
        val items = page.items.map { PageItem(it.id, decode(it.value)) }
        pending.remove(page.id)
        session?.resolve(id, items, page.total)
    }
    fun reject(request: PageCancellation) {
        pending.remove(request.id)?.let { session?.reject(it) }
    }
    fun close() {
        session?.close()
        session = null
        pending.clear()
    }
}
